packages/services/src/state_listener.rs

```diff
@@ -472,4 +538,238 @@ mod tests {
 
         Ok(())
     }
+
+    #[tokio::test]
+    async fn fine_to_have_nothing_to_check() -> Result<()> {
+        // given
+        let setup = test_utils::Setup::init().await;
+
+        let mut listener = StateListener::new(
+            ports::l1::MockApi::new(),
+            setup.db(),
+            5,
+            TestClock::default(),
+            IntGauge::new("test", "test").unwrap(),
+        );
+
+        // when
+        let res = listener.run().await;
+
+        // then
+        assert!(res.is_ok());
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn a_pending_tx_got_squeezed_out() -> Result<()> {
+        // given
+        let setup = test_utils::Setup::init().await;
+        let _ = setup.insert_fragments(0, 1).await;
+
+        let tx_hash = [0; 32];
+        setup.send_fragments(tx_hash).await;
+        let mut l1 = ports::l1::MockApi::new();
+        l1.expect_get_block_number()
+            .returning(|| Box::pin(async { Ok(5.into()) }));
+
+        l1.expect_get_transaction_response()
+            .with(eq(tx_hash))
+            .once()
+            .return_once(|_| Box::pin(async { Ok(None) }));
+
+        l1.expect_is_squeezed_out()
+            .with(eq(tx_hash))
+            .once()
+            .return_once(|_| Box::pin(async { Ok(true) }));
+
+        let mut sut = StateListener::new(
+            l1,
+            setup.db(),
+            5,
+            TestClock::default(),
+            IntGauge::new("test", "test").unwrap(),
+        );
+
+        // when
+        sut.run().await?;
+
+        // then
+        assert!(!setup.db().has_pending_txs().await?);
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn block_inclusion_of_replacement_leaves_no_pending_txs() -> Result<()> {
+        // given
+        let setup = test_utils::Setup::init().await;
+
+        // Insert multiple fragments with the same nonce
+        let _ = setup.insert_fragments(0, 1).await;
+
+        let start_time = Utc::now();
+        let orig_tx_hash = [0; 32];
+        let orig_tx = L1Tx {
+            hash: orig_tx_hash,
+            created_at: Some(start_time),
+            ..Default::default()
+        };
+
+        let replacement_tx_time = start_time + Duration::from_secs(1);
+        let replacement_tx_hash = [1; 32];
+        let replacement_tx = L1Tx {
+            hash: replacement_tx_hash,
+            created_at: Some(replacement_tx_time),
+            ..Default::default()
+        };
```

Comment:
being explicit here about replacements requiring the same nonce might save us some trouble later